build_runner/lib/src/asset_graph/graph.dart

```diff
@@ -94,38 +100,79 @@ class AssetGraph {
   /// set of assets that need to be deleted.
   Iterable<AssetId> updateAndInvalidate(Map<AssetId, ChangeType> updates) {
     var deletes = new Set<AssetId>();
-    var seen = new Set<AssetId>();
     void clearNodeAndDeps(AssetId id, ChangeType rootChangeType,
-        {AssetId parent}) {
-      if (seen.contains(id)) return;
-      seen.add(id);
+        {AssetId parent, bool rootIsSource}) {
```

Comment:
When is the root not an original source? We should not be reporting file change events for generated sources afaik.